### DISCUSSION TITLE
Trying to use UBL?

### DISCUSSION THREAD

**Discussion by synman** [synman/Ender-3-S1-Pro-Firmware#23] [Features]
If you've never used it before, its syntax is a bit different than ABL.  Read all about it in the Marlin docs.

It is fully integrated into the touchscreen if you flash the UBL build and will produce a 10x10 grid, 100 points, spread across every point of the bed that the probe can reach.

Will keep this open to answer any questions that arise and share some experiences I've had with it in follow up comments.

<img width="798" alt="Screenshot 2023-02-03 at 9 29 16 PM" src="https://user-images.githubusercontent.com/1299716/216741756-b5a4ecf8-c5ab-4e5b-9181-0cbe90ce4a5b.png">

**Comment by synman** [comment]
A quick walkthrough of the UBL workflow on the S1 Pro / Plus:

1. Mesh generation starts with a `G29 P1` command.  The probe will run through the grid, probing each of the 100 points along the way.  Running `G29 P1 T V4` will provide an updated topology report per probe as well as some diagnostic data as it progresses.
```
Recv: Bed Topography Report:
Recv: 
Recv:     (  0,183)                                                              (203,183)
Recv:         0       1       2       3       4       5       6       7       8       9
Recv:  9 | +0.05  -0.01  -0.07  -0.09  -0.09  -0.03  -0.02  +0.01  +0.04  +0.13
Recv:    |
Recv:  8 | +0.05  +0.01  -0.04  -0.07  -0.03  -0.03  -0.01  +0.03  +0.07  +0.14
Recv:    |
Recv:  7 | +0.05  +0.01  -0.04  -0.05  -0.05  -0.05 [-0.02] +0.01  +0.03  +0.11
Recv:    |
Recv:  6 | +0.05  +0.01  -0.04  -0.06  -0.06  -0.06  -0.04  -0.02  +0.02  +0.08
Recv:    |
Recv:  5 | +0.04  -0.01  -0.05  -0.07  -0.07  -0.08  -0.06  -0.02  +0.02  +0.09
Recv:    |
Recv:  4 | +0.05  -0.01  -0.06  -0.08  -0.08  -0.08  -0.06  -0.03  -0.01  +0.06
Recv:    |
Recv:  3 | +0.05  -0.02  -0.07  -0.10  -0.09  -0.09  -0.08  -0.05  -0.02  +0.04
Recv:    |
Recv:  2 | +0.03  -0.05  -0.12  -0.13  -0.11  -0.10  -0.10  -0.08  -0.06  +0.00
Recv:    |
Recv:  1 | -0.01  -0.09  -0.13  -0.16  -0.14  -0.11  -0.09  -0.06  -0.05  +0.03
Recv:    |
Recv:  0 | +0.06  -0.01  -0.06  -0.09  -0.06  -0.05  -0.03  +0.01  +0.05  +0.13
Recv:         0       1       2       3       4       5       6       7       8       9
Recv:     (  0,  0)                                                              (203,  0)
```
2. The probe may fail one or more times while running through the 100 points.  Let it get to the end and we'll follow up with a Continue command to finish mesh population.  Once the initial P1 session is complete, issue a Continue like so:  `G29 P1 C T V4`

3. Now that we have a full mesh, execute the Fill Gaps command like so:  `G29 P3`

4. You should save your mesh at this point with the Save command: `G29 S1`

5. You can activate your mesh (this should also be in your slicer's startup gcode) with `M420 L1 S1 V` 

If you're using the Touchscreen, all of this is built into it with the exception of step 5 :)

G29 UBL is documented here:  https://marlinfw.org/docs/gcode/G029-ubl.html

**Reply by JohnScottUK** [reply]
I tried UBL 100 but it ended up trying to level on my clips holding the glass bed down. So I went back to ubl25.

**Reply by synman** [reply]
So long as it isn't risking any harm to your probe, you could  invalidate those points and `G29 P3` them.

I hate the idea of breaking a probe though so please don't try this if there's a chance it could damage things.

I have been having a lot of success with the 100 point mesh and tilting (G29 J).  I imagine the 25 pt one will do good as well but possibly miss a couple nooks and crannies.

**Comment by synman** [comment]
After you've set your Z-Offset, if you want to do actual printed validation of your work, you can do so by printing a Mesh Validation Grid (https://marlinfw.org/docs/gcode/G026.html)

The following example set an extruder temperature of 205C, nozzle size of .6mm at .2mm layer height.  It will generate a 100 point grid that visually matches the mesh you generated above.
```
G26 S.6 R100 L.2 O0 C 
```
![PXL_20230205_221150635](https://user-images.githubusercontent.com/1299716/216851098-4d22fee1-093d-4553-9d76-1ea17c16be90.jpg)

G26 also works with ABL based meshes.  For example, to produce a printed grid if using the ABL version of the firmware, specify `25` as the `R` parameter value instead of `100`.

**Comment by synman** [comment]
A really useful command is Mesh point Invalidation.  Here are a couple examples of it in use:

`G29 I T` will delete the mesh point closest to the nozzle's current position.

`G29 I9 T` will delete the 9 closest mesh points to the nozzle's current position.

Executing Continue (`G29 C`) after deleting mesh points will have the machine re-probe them.

Here's an example of clearing all 4 corners of your mesh and re-probing them.  This is useful if you are working on tramming your bed:
```
G29 I X235 Y0
G29 I X235 Y225
G29 I X0 Y225
G29 I X0 Y0
G29 T
G29 P1 C T V4
G29 S
```

**Comment by synman** [comment]
There are now two variations of UBL available.

_ubl25 is a 5x5 (25 point) mesh
_ubl100 is a 10x10 (100 point) mesh

**Comment by synman** [comment]
Another super cool feature of UBL is the ability to shift an existing mesh on its axes based on re-sampling 3 or more points of the bed.

The premise is that the only part of a mesh that can change is the orientation of it at its corners (the knobs).   As part of your normal printing workflow, you can sample 3 or 4 or more points of the bed to determine if the corners have shifted, and if so rebalance the entire mesh to match this variation!

UBL literally compares the two planes (the plane of the nozzle and the plane of the bed) and with as little as 3 points of data can shift the entire mesh to match the detected variance.

`M29 J` is the command you want to use.

Here are some examples of it in use:
```
M29 J ; re-sample and shift front corner / front center / front corner points
M29 J2 ; re-sample all 4 corners
M29 J3 ; re-sample all 4 corners, their center points, and Y center of the bed
```
Basic rule here is the value of J gets squared and that is how many points are re-sampled.

2 x 2 = 4 points
3 X 3 = 9 points
4 x 4 = 16 points
etc

The underlying mesh does not get saved with this command.  The approach I take is to keep the original mesh intact and just reapply the shift every time I start a job with some custom gcode in my slicer startup script:
```
G28 ; home all axis
;
M420 L1 V ; use saved mesh
G29 J2 T; shift mesh (4 corners)
M420 S1 V; activate mesh
```

* NOTE - this is NOT new information or functionality.  I'm merely redescrbiing what many before me have already done.

**Comment by synman** [comment]
A useful setting if using Octoprint and getting disconnected due to probe failures:

![Screenshot_20230221-002218.png](https://user-images.githubusercontent.com/1299716/220255254-bbe5526b-b2d9-412d-a467-75d0ebba1b02.png)

**Comment by MrDB42** [comment]
If you are frustrated and tired of dealing with failed points (locking up Octoprint or even the printer itself) use the E option [Stow probe after probing each point], with any G29 P1 commands. i.e.
`G29 P1 E`
or 
`G29 P1 T V4 E`
or
`G29 P1 C T V4 E`
etc.

Unfortunately, this will add more time to an already lengthy process. But I found it to be worth it, for not dealing with failures.

My hypothesis is that the CR-Probe has some protection circuitry built in and having it extended for a prolonged period it will retract the probe. It is an electromagnet after all that engages the probe, and it will heat up. 

I thought my probe was at issue, so I ordered and installed a new one, only to find out that it too would fail repeatedly and consistently at the same points my old probe did. (It wasn't a total waste of money by-the-way, I noticed that one of the screw tabs was cracked on my old probe. I must have overtightened the screw at one time when experimenting with Zero-Y offset mounts).

**Reply by synman** [reply]
I too played with the stow command.   It is not a silver bullet.  Enabling it does change the behavior a bit though.

**Reply by MrDB42** [reply]
True, it may not be a silver bullet. But so far, I can finally try it out. Previously to finding the E option, it didn't matter if I disconnected Octoprint and just ran it via the touchscreen, it would error out completely. The printer itself would halt on "Probe failed" and the only way I could get out of it was to power cycle.

At least this way I can play around with it.

**Reply by MrDB42** [reply]
Oh, another way to tram your bed is to use the G30 command over your adjusting screws (or as close as you can get to them). Just verify your probe offset prior (my probe is actually at X-32 Y-40), AND turn off bedleveling while doing so.

So I probe and adjust the following position order:
```
G30 X25 Y28
G30 X198 Y185
G30 X25 Y185
G30 X198 Y28
```
(Adjusting via diagonals, goes quicker than just going around in a circle, but it still takes time. And this why I asked for the increased decimal display)

**Reply by synman** [reply]
> My hypothesis is that the CR-Probe has some protection circuitry built in and having it extended for a prolonged period it will retract the probe. It is an electromagnet after all that engages the probe, and it will heat up.

When I was originally getting everything functional for UBL, I noticed a situation where the probe would always fail at the same relative point, regardless of if it was doing 25 points or 100 points.  What I ultimately figured out was that I had something screwed up from a bed size / offsets perspective and UBL was getting cranky about the commanded position not being within what it considered the appropriate "bounds".

We're beyond that as I had addressed it prior to releasing the feature.  Often times my 2-3 failures on a 100 pt run are for a cluster of points all within the center of the bed, for example, but I figured I should share this (painful) learning on my part as I was working out all the nuances.

I should also try some further testing with a 25 point grid.   I've been hesitant to mess much with it for similar reasons... the time it takes to get a good mesh and maintain it.

I'm tempted to do some testing with ABL with 50/75/100 mesh points to see what happens.   I have this suspicion it would be okay, possibly because `G29_RETRY_AND_RECOVER` is supported with ABL but not UBL.

**Comment by MrDB42** [comment]
Mine is always point 27, 54, 82. Even with an M48 - Probe Repeatability Test (example: `M48 P50 X110 Y110 V2`), when it hits number 27 it errors out. So that kind of points to the probe itself (in my mind).

**Reply by synman** [reply]
I hadn't bothered to test it with M48 but suggested it be tried in another thread.   Very interesting!  I'm half tempted to re-engage on this since it can be recreated with it.   Makes it easier to recreate without having to blow away my mesh.

**Reply by synman** [reply]
nice... we're seeing the exact same thing:

```
Recv: //action:notification M48 Point: 27/50
[...]
Recv: //action:notification Probing Failed
Recv: Error:Probing Failed
```

**Reply by synman** [reply]
I'm rerunning it now with retract....  curious to see how this ends:

```
M48 P50 X110 Y110 V2 E1
```

**Reply by lipa46** [reply]
I can confirm that during bad leveling my probe fails at the same points as of @MrDB42 maybe +/-1

**Reply by synman** [reply]
@MrDB42 dang you're good!

just made it all the way thru the 50 point repeatability test with retract enabled.

```
Recv: 50 of 50: z: 0.009
Recv: Finished!
Recv: Mean: 0.017009 Min: 0.008 Max: 0.024 Range: 0.016
Recv: Standard Deviation: 0.005727
```

**Reply by synman** [reply]
> I can confirm that during bad leveling my probe fails at the same points as of @MrDB42 maybe +/-1

ah yes, @lipa46 you're the one I asked to run the repeatability test.  you were also reporting much broader issues though.   Have you been able to narrow it down to these specific conditions?

**Reply by lipa46** [reply]
I have run the repeatability test and at 10 repetitions and the standard deviation of 0.003 something and range of 0.014. Which what I found on the Internet should be well in specification. As you run your last repeatability test was you bed hot? It is posable what @MrDB42 pointed out that the sensor could have some protection. During probing as the probe is metal and it comes in contact with hot bed it can overheat much quicker.

**Reply by synman** [reply]
we're not running the repeatability test to measure the accuracy of the probe here... we're running it in hopes it causes the probe to outright fail.

**Reply by lipa46** [reply]
Yes I understand that. In current can't test it. But try heating up the bad and re run the 50 cycle test.

**Reply by MrDB42** [reply]
@lipa46 
 Oh, try this:
`M48 P50 X110 Y110 V2 E`
Usually, it will error out at number 27. We just figured that stowing the probe each time overcomes this issue. Heating the bed makes no difference actually for what we are testing. But, true in reality, we should be heating the bed when doing ACTUAL repeatablity testing.

**Reply by synman** [reply]
there is a post I read recently about electrical interference wreaking havoc on certain S1s with the CR touch.   I haven't made the leap to that here yet, but it is in the back of my mind.   Apparently for some people it so bad that probe fails 99% of the time.   In that post, they ran everything to a common ground... the bed, the probe, the heaters / etc and in doing so were able to resolve their issues.

**Reply by MrDB42** [reply]
That IS a possibility. Electromagnetic interference is a bloody nightmare (hence the UL testing which I'm pretty darn sure Creality hasn't done). I **really** don't want to go down that road. Just find a decent work-around that works consistently. 

But I'm sure we aren't the first ones to run into this issue, otherwise the "E" option wouldn't be there.

**Reply by synman** [reply]
found the post...  its on the Creality Ender 3 Users discord server:

<img width="1048" alt="Screenshot 2023-02-27 at 5 00 44 PM" src="https://user-images.githubusercontent.com/1299716/221695893-090ac83f-bdc8-4577-a7d0-892b3b784750.png">

**Reply by synman** [reply]
I do like the common ground idea... I'm actually surprised (even though I shouldn't be) that we don't already have a common ground between the PSU and the chassis.  I'm not in hurry to crack open my case, but it is something to consider.   In the meantime,  I'm going to see if I can "slipstream" in a periodic retract within probe.c.... let's say retract once every 20 probes while in a probe loop :)

**Reply by MrDB42** [reply]
That makes perfect sense for those who have added the CR-Touch as an afterthought. You need to ground out or choke somewhere. I haven't broken out the multimeter on the S1 Pro to verify. But I'm pretty sure that PCB which the CR-Touch connects to is properly grounded.

Oh, I have opened things up in the past and DID note that PSU is properly grounded.

**Reply by MrDB42** [reply]
Oh god, now you've got me wondering if I was looking at the right place for the PSU grounding. I know that the bed is grounded properly. I guess I'm going diving into it's guts now.

**Reply by synman** [reply]
sorry, I did not mean to open that can of worms.    I'm going the software route at the moment.... enabling debugging to determine exactly where this thing is failing.

**Reply by synman** [reply]
here's what I've got to work with:
```
Recv: //action:notification M48 Point: 27/50
Recv: >>> Probe::probe_at_point  X141.5000 Y151.8000 Z8.0315
Recv: ...(110.0000, 110.0000, raise, 0, probe_relative)
Recv:   current_position= X141.5000 Y151.8000 Z8.0315 :
Recv: >>> do_blocking_move_to  X141.5000 Y151.8000 Z8.0315
Recv: >  X141.5000 Y151.8000 Z8.0315
Recv: <<< do_blocking_move_to  X141.5000 Y151.8000 Z8.0315
Recv:   current_position= X141.5000 Y151.8000 Z8.0315 : Probe::set_deployed
Recv: deploy: 1
Recv: >>> Probe::run_z_probe  X141.5000 Y151.8000 Z8.0315
Recv: >>> Probe::probe_down_to_z  X141.5000 Y151.8000 Z8.0315
Recv: >>> do_blocking_move_to  X141.5000 Y151.8000 Z8.0315
Recv: >  X141.5000 Y151.8000 Z1.0000
[...]
Recv: <<< do_blocking_move_to  X141.5000 Y151.8000 Z1.0000
Recv:   current_position= X141.5000 Y151.8000 Z1.0000 : sync_plan_position
Recv: <<< Probe::probe_down_to_z  X141.5000 Y151.8000 Z1.0000
Recv: SLOW Probe fail! - No trigger.
Recv: <<< Probe::run_z_probe  X141.5000 Y151.8000 Z1.0000
Recv:   current_position= X141.5000 Y151.8000 Z1.0000 : Probe::set_deployed
Recv: deploy: 0
Recv: Probe::do_z_raise(10.0000)
Recv: >>> do_blocking_move_to  X141.5000 Y151.8000 Z1.0000
Recv: >  X141.5000 Y151.8000 Z13.0000
[...]
Recv: <<< do_blocking_move_to  X141.5000 Y151.8000 Z13.0000
Recv: BLTouch STOW requested
Recv: BLTouch Command :90
Recv: bltouch.stow_proc() end
Recv: >>> do_blocking_move_to  X141.5000 Y151.8000 Z13.0000
Recv: >  X141.5000 Y151.8000 Z13.0000
Recv: <<< do_blocking_move_to  X141.5000 Y151.8000 Z13.0000
Recv: //action:notification Probing Failed
Recv: Error:Probing Failed
Recv: measured_z: 0.0000
Recv: <<< Probe::probe_at_point  X141.5000 Y151.8000 Z13.0000
Recv:   current_position= X141.5000 Y151.8000 Z13.0000 : Probe::set_deployed
Recv: deploy: 0
Recv: >>> set_bed_leveling_enabled  X141.5000 Y151.8000 Z13.0000
Recv: <<< set_bed_leveling_enabled  X141.5000 Y151.8000 Z13.0000
[...]
Recv: ok
[...]
```

**Reply by synman** [reply]
I'm talking with one of the core marlin devs and he just gave me a couple pointers.   I may be blowing out the heap/stack.  he also suggested I revert a change I made for multiple probing.

**Reply by MrDB42** [reply]
Hmmm.... Interesting. I think we can rule out hardware issues then.

BTW, I just opened up the guts of my S1 Pro, and I can confirm that EVERYTHING is properly grounded. ALL the way to the _tip of the nozzle_ (wink wink nudge nudge).

**Reply by synman** [reply]
so both recommendations did not pan out.....  memory checks out and changing multiple probing merely extended the run a couple more probing events.

I did come up with a workaround.   a patch of sorts directly to G29_ubl.cpp while in the probing loop:
```
      // override for stow_probe
      const bool stowing = stow_probe || point_num % 10 == 0;
```
basically force a stow / retract every 10 probes

**Reply by synman** [reply]
this fix (along with an upstream merge and an unrelated baby stepping issue fix) are going live shortly.  just waiting for the build to finish

**Reply by MrDB42** [reply]
Stowing the probe every 10 probes works for me. I was thinking about suggesting that earlier, but I wanted to see if the suggestions from the Marlin developer panned out also.

I just hope that you didn't forget to comment your code as to why you put that in.

**Reply by synman** [reply]
I declared it as CRTOUCH_STOW_HACK lol.

I heard back from him again this morning.  Once I described the behavior in detail, he said yeah, par for the course.

**Reply by MrDB42** [reply]
Hahahahaha!! Most excellent. 

I used to not document my code very well (if at all). Until one of my professors got tired of it, and failed every assignment I turned in. Then I went nuts documenting everything. 5 lines of code turned into a paragraph or two of comments explaining everything in detail. He LOVED it (I thought I was going to be chastised for too much commenting). But let me tell you, that habit was a life saver later on in life when having to go back and fix things.

But I'm still amazed at how accurate the probe is, for an electromagnet plunger and HALL effect sensors.

**Reply by synman** [reply]
I've got one more thing to try.  It was suggested to outright disable fast probing.  I'll try that next and weigh it against the hack.

The word safety was used... At this point though what we've got is no less safe than what Creality delivered.

**Reply by MrDB42** [reply]
Personally, I think the stow option is the safest and easiest. The only other thing I can think of is to increase from 10 to 20 before stowing, just to speed up the process. 

I have a gut feeling that fail on probe 27 is consistent across all the CR-Probes. At least that's what I would do if I was in charge of designing the probe, to protect itself from maniacs, like us, who like to do 100 bed probe points.

And why would disabling fast probing be not as safe? Wouldn't you just be doing slow probing otherwise? And if that is the case, then I have a feeling that the probe would stow on its own even sooner.

**Reply by synman** [reply]
I've been playing a little.

Checkout out `M401 H` and `M401 S0/1`

You can turn off/on fast probe via software.

Of course, I have to revert the hack to properly test it.  I think I'm going to revert it from M48 so we can play more.

There is a suspicion being floated there the S1 models have a newer generation version of the CR touch and it's only a matter of time before people start replacing them with older, far less reliable, ones already on shelves.

**Reply by MrDB42** [reply]
Hmm. Interesting learn something new every day. 

But now I understand why doing a slow probe might be a bad thing. The print head _could_ be moving faster than the probe deploy in slow mode. Which I can see someone doing that, maybe.

**Reply by synman** [reply]
So we're definitely at a fork in the road.

With fast probe disabled via `M401 H S0` I'm able to get through all 50 M48 probes without issue (I backed out my stow hack to test this).   M401 H S1/0 is also an EEPROM based register so M500 will persist it across boots.

I did some basic benchmarking and fast probe is almost twice as fast as slow probe (about 65% of the time it takes to do a slow probe).

Here's what I'm thinking:

Leave `CRTOUCH_STOW_HACK` in place on `G29` but only activate it when `BLTOUCH_HS_MODE` is active.  If high speed mode is disabled via `M401 H S0`, there's no point having the stow hack active too.

Since no one has a CR Touch that works 100% of the time, this feels like the best supportable approach with some tweaks.  I'm going to decrease the stow hack frequency to once every 20 probes and if someone experiences problems with it at that rate then they can disable high speed mode entirely.

**Reply by synman** [reply]
I'll also be dropping `MULTIPLE_PROBING` back down to the Creality default of 2 instead of 3.  I've heard from multiple people now, there's no real benefit of the extra probe, so it's outta here.

**Reply by MrDB42** [reply]
Now THAT'S creative. "Mikey Likes It!" (Yes, I'm a borderline boomer). I thought you increased the probe count but wasn't sure. I too don't really see the need to have a 3-probe count. If it can't hit the mark on 2, then you have serious issue with the probe. It is a really simple device.

So, in summary...
If Fast Probe enabled, then G29 stow after 20 probes. Else if Slow Probe enabled then G29 no stow.
Caveats is that if Slow Probe enabled then 2x time is needed for G29 to process.

Sounds like a workable solution to me.

One other 'gotcha' to look at: "M48 - Probe Repeatability Test". Can the same be implemented with this command?

**Reply by synman** [reply]
I have removed `CRTOUCH_STOW_HACK` from M48.  I figure we need to keep it true to what it is meant for.  I hadn't considered adding a parameter to it for the hack but I guess I could.  Heck, I could just allow you to specify what interval.

Totally down for Life cereal.   You may have me by a couple years but I'm getting up there myself.... I'm a couple years past the 50 mark.

**Reply by MrDB42** [reply]
You know skip the M48 hack. Keep it true.

Dang, I think we might be the same - 54 here

**Reply by synman** [reply]
you got me by 2!

**Comment by MrDB42** [comment]
@lipa46 Yeah, I'm not fully confident about points after number 27. But up to number 26 works fine, for sure.
@synman Thank you! I to just ran an M48 with "E" option also, and I too made it all the way through finally.

BTW. What bed visualization plug-in do you use, out of curiosity. I use BedVisualizer, and man, my meshes come out "blocky".
![bedmesh](https://user-images.githubusercontent.com/122186614/221690297-1a744d11-2b98-4c4e-b2a0-978f1e2199c7.png)

**Reply by synman** [reply]
I too use jneilliii's bed visualizer plugin.   He's a great guy.  I lurk with him in the Octoprint discord server quite often as we're both considered maintainers / plugin developers.

Here's a little trick that was imparted on me:

<img width="655" alt="Screenshot 2023-02-27 at 4 29 33 PM" src="https://user-images.githubusercontent.com/1299716/221690070-ee3c6cfe-8a72-49b3-ac80-196a410696b1.png">

<img width="882" alt="Screenshot 2023-02-27 at 4 30 29 PM" src="https://user-images.githubusercontent.com/1299716/221690150-be285e5a-3af0-4f66-bc2e-3e42b687dda0.png">

Be careful though... if you're remotely obsessive-compulsive (like me) it can be quite a rabbit hole because the auto,auto settings keep you chasing the perfect mesh.

**Reply by MrDB42** [reply]
Oh god... don't talk to me about obsessive-compulsive issues. 3D printers are NOT good for people who have obsessive-compulsive issues.

**Reply by synman** [reply]
A .2mm variance is not bad at all.

**Reply by MrDB42** [reply]
I was much WORSE. I lapped the bed on a piece of glass (not fun). I'm still thinking about outsourcing a custom bed at 4-5mm thick properly machined to a decent level of parallelism and  smoothness (my micro mill just isn't big enough to allow me to do it myself).
